### ISSUE TITLE
Hang while loading roslyn.sln

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#40404]
Opened up Roslyn.sln and hit a hang during this.  Uploading dump now.

**Comment by CyrusNajmabadi** [comment]
Dump is here: https://drive.google.com/file/d/1xd2M6mWpTNTzrAc3QK-AvxzbncsZX_RT/view?usp=sharing

Tagging @jinujoseph @sharwell @mavasani

**Comment by mavasani** [comment]
Tagging @lifengl in case there is an easy way to detect the underlying reason for the deadlock with **CpsJoinableTaskContext.OnHangDetected** on top of the main thread stack.

Tagging @tmat as the dump has large number of threads (300+) disposing the remote host client `Microsoft.CodeAnalysis.Remote.RemoteHostClient.Connection.Dispose()`.

## Main thread

```
Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.StaThreadMarshaler.CpsJoinableTaskContext.OnHangDetected(Microsoft.VisualStudio.Threading.JoinableTaskContext.HangDetails hangDetails)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskContext.OnHangDetected(System.TimeSpan hangDuration, int notificationCount, System.Guid hangId)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronouslyCore(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.DelegatingJoinableTaskFactory.WaitSynchronously(System.Threading.Tasks.Task task)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.CompleteOnCurrentThread()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask.Join(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTask<object>.Join(System.Threading.CancellationToken cancellationToken)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__3()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsTaskLibraryHelper.AsVsTask.AnonymousMethod__1(object sender, Microsoft.VisualStudio.Shell.BlockingTaskEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.RaiseOnMarkedAsBlockingEvent(Microsoft.VisualStudio.Shell.Interop.IVsTask blockedTask)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog<System.__Canon>(System.Func<System.__Canon> function)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InvokeWithWaitDialog(System.Action action)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.InternalGetResult(bool ignoreUIThreadCheck)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.Services.VsTask.GetResult()	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.GetProvisionalViewingStatusAsync(Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.SimpleHierarchyNode node)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.GetPropertySlow.AnonymousMethod__18()	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.ExecuteJob<Microsoft.VisualStudio.Threading.EmptyStruct>(System.Func<System.Threading.Tasks.Task> asyncMethod, Microsoft.VisualStudio.Threading.JoinableTask job)	Unknown
 	Microsoft.VisualStudio.Threading.dll!Microsoft.VisualStudio.Threading.JoinableTaskFactory.RunAsync(System.Func<System.Threading.Tasks.Task> asyncMethod, bool synchronouslyBlocking, Microsoft.VisualStudio.Threading.JoinableTaskCreationOptions creationOptions, System.Delegate entrypointOverride)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.ProjectMultiThreadedService.ExecuteSynchronously(System.Func<System.Threading.Tasks.Task> asyncAction)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.GetPropertySlow.AnonymousMethod__0()	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.dll!Microsoft.VisualStudio.ProjectSystem.VS.HResult.Invoke(System.Func<Microsoft.VisualStudio.ProjectSystem.VS.HResult> action, System.IServiceProvider vsShellServiceProvider, Microsoft.VisualStudio.ProjectSystem.IProjectFaultHandlerService projectFaultHandlerService, Microsoft.VisualStudio.ProjectSystem.UnconfiguredProject project)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.GetPropertySlow(uint itemId, int propId, out object propValParam)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNode.GetProperty(uint itemId, int propId, out object propVal)	Unknown
 	Microsoft.VisualStudio.ProjectSystem.VS.Implementation.dll!Microsoft.VisualStudio.ProjectSystem.VS.Implementation.Package.ProjectNodeWrapper.GetProperty(uint itemid, int propid, out object pvar)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.HierarchyUtilities.TryGetHierarchyProperty.AnonymousMethod__0()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.ErrorHandler.CallWithCOMConvention(System.Func<int> method, bool reportError, bool setShellErrorInfo)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.HierarchyUtilities.TryGetHierarchyProperty<uint>(Microsoft.VisualStudio.Shell.Interop.IVsHierarchy hierarchy, uint itemid, int propid, System.Func<object, uint> converter, out uint value)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.VisualStudio.Shell.VsShellUtilities.GetProvisionalViewingStatus(Microsoft.VisualStudio.Shell.Interop.IVsHierarchy hierarchy, uint itemid)	Unknown
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.HierarchyItem.ProvisionalViewingStatus.get()	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.PivotTreeView.Microsoft.Internal.VisualStudio.PlatformUI.ISelectionPreviewController.CanPreviewItem(object item)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.PivotTreeView.CanPreviewSelection(System.Windows.Controls.SelectionChangedEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.UI.Internal.dll!Microsoft.VisualStudio.PlatformUI.HierarchyPivotTreeView.CanPreviewSelection(System.Windows.Controls.SelectionChangedEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.SelectionPreview.PreviewItemState.DelayedPreviewSelection(System.Windows.Controls.SelectionChangedEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.SelectionPreview.PreviewItemState.OnSelectionChanged(object sender, System.Windows.Controls.SelectionChangedEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.SelectionChangedEventArgs.InvokeEventHandler(System.Delegate genericHandler, object genericTarget)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventArgs.InvokeHandler(System.Delegate handler, object target)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, System.Windows.RoutedEventArgs routedEventArgs)	Unknown
 	PresentationCore.dll!System.Windows.EventRoute.InvokeHandlersImpl(object source, System.Windows.RoutedEventArgs args, bool reRaised)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.RaiseEventImpl(System.Windows.DependencyObject sender, System.Windows.RoutedEventArgs args)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.RaiseEvent(System.Windows.RoutedEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ListBox.OnSelectionChanged(System.Windows.Controls.SelectionChangedEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.Primitives.Selector.InvokeSelectionChanged(System.Collections.Generic.List<System.Windows.Controls.ItemsControl.ItemInfo> unselectedInfos, System.Collections.Generic.List<System.Windows.Controls.ItemsControl.ItemInfo> selectedInfos)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.Primitives.Selector.SelectionChanger.End()	Unknown
 	PresentationFramework.dll!System.Windows.Controls.Primitives.Selector.SelectionChanger.SelectJustThisItem(System.Windows.Controls.ItemsControl.ItemInfo info, bool assumeInItemsCollection)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ListBox.MakeSingleSelection(System.Windows.Controls.ListBoxItem listItem)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ListBox.NotifyListItemClicked(System.Windows.Controls.ListBoxItem item, System.Windows.Input.MouseButton mouseButton)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ListBoxItem.HandleMouseButtonDown(System.Windows.Input.MouseButton mouseButton)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ListBoxItem.OnMouseLeftButtonDown(System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.DraggableListBoxItem.OnMouseLeftButtonDown(System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.VirtualizingTreeViewItem.OnMouseLeftButtonDown(System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.15.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.PivotTreeViewItem.OnMouseLeftButtonDown(System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.OnMouseLeftButtonDownThunk(object sender, System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	PresentationCore.dll!System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(System.Delegate genericHandler, object genericTarget)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventArgs.InvokeHandler(System.Delegate handler, object target)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, System.Windows.RoutedEventArgs routedEventArgs)	Unknown
 	PresentationCore.dll!System.Windows.EventRoute.InvokeHandlersImpl(object source, System.Windows.RoutedEventArgs args, bool reRaised)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.ReRaiseEventAs(System.Windows.DependencyObject sender, System.Windows.RoutedEventArgs args, System.Windows.RoutedEvent newEvent)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.OnMouseDownThunk(object sender, System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	PresentationCore.dll!System.Windows.Input.MouseButtonEventArgs.InvokeEventHandler(System.Delegate genericHandler, object genericTarget)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventArgs.InvokeHandler(System.Delegate handler, object target)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, System.Windows.RoutedEventArgs routedEventArgs)	Unknown
 	PresentationCore.dll!System.Windows.EventRoute.InvokeHandlersImpl(object source, System.Windows.RoutedEventArgs args, bool reRaised)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.RaiseEventImpl(System.Windows.DependencyObject sender, System.Windows.RoutedEventArgs args)	Unknown
 	PresentationCore.dll!System.Windows.ContentElement.RaiseTrustedEvent(System.Windows.RoutedEventArgs args)	Unknown
 	PresentationCore.dll!System.Windows.Input.InputManager.ProcessStagingArea()	Unknown
 	PresentationCore.dll!System.Windows.Input.InputManager.ProcessInput(System.Windows.Input.InputEventArgs input)	Unknown
 	PresentationCore.dll!System.Windows.Input.InputProviderSite.ReportInput(System.Windows.Input.InputReport inputReport)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndMouseInputProvider.ReportInput(System.IntPtr hwnd, System.Windows.Input.InputMode mode, int timestamp, System.Windows.Input.RawMouseActions actions, int x, int y, int wheel)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndMouseInputProvider.FilterMessage(System.IntPtr hwnd, MS.Internal.Interop.WindowMessage msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndSource.InputFilterMessage(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndWrapper.WndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.SubclassWndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam)	Unknown
```

## Threads disposing the remote host client
```
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout, bool exitContext) Line 203	C#
 	mscorlib.dll!System.Threading.Monitor.Wait(object obj, int millisecondsTimeout) Line 213	C#
 	mscorlib.dll!System.Threading.ManualResetEventSlim.Wait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 688	C#
 	mscorlib.dll!System.Threading.Tasks.Task.SpinThenBlockingWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3320	C#
 	mscorlib.dll!System.Threading.Tasks.Task.InternalWait(int millisecondsTimeout, System.Threading.CancellationToken cancellationToken) Line 3259	C#
 	mscorlib.dll!System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task task) Line 153	C#
 	[Waiting on Async Operation, Open Tasks View in Parallel Stacks Window to view Async Call Stacks]	
 	StreamJsonRpc.dll!StreamJsonRpc.MessageHandlerBase.Dispose()	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.JsonRpc.OnJsonRpcDisconnected(StreamJsonRpc.JsonRpcDisconnectedEventArgs eventArgs)	Unknown
 	StreamJsonRpc.dll!StreamJsonRpc.JsonRpc.Dispose(bool disposing)	Unknown
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.JsonRpcConnection.ServiceJsonRpcEx.Dispose(bool disposing) Line 88	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.JsonRpcConnection.Dispose(bool disposing) Line 60	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClient.Connection.Dispose() Line 146	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.ConnectionManager.Free(string serviceName, Microsoft.VisualStudio.LanguageServices.Remote.JsonRpcConnection connection) Line 146	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Remote.ServiceHubRemoteHostClient.ConnectionManager.PooledConnection.Dispose(bool disposing) Line 53	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClient.Connection.Dispose() Line 146	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.SessionWithSolution.CreateAsync(Microsoft.CodeAnalysis.Remote.RemoteHostClient.Connection connection, Microsoft.CodeAnalysis.Solution solution, System.Threading.CancellationToken cancellationToken) Line 47	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Remote.RemoteHostClientExtensions.TryCreateSessionAsync(Microsoft.CodeAnalysis.Remote.RemoteHostClient client, string serviceName, Microsoft.CodeAnalysis.Solution solution, object callbackTarget, System.Threading.CancellationToken cancellationToken) Line 53	C#
```

**Comment by CyrusNajmabadi** [comment]
@tmat Do you know what's up here?

**Comment by mavasani** [comment]
I suspect this is the same as https://github.com/dotnet/roslyn/issues/40455#issuecomment-566851545. @CyrusNajmabadi I think you may want to move back to 16.4.2 until 16.5 Preview2 is available.

**Comment by tmat** [comment]
> Tagging @tmat as the dump has large number of threads (300+) disposing the remote host client Microsoft.CodeAnalysis.Remote.RemoteHostClient.Connection.Dispose().

The threads are hanging in JSON RPC [MessageHandlerBase.Dispose](https://github.com/microsoft/vs-streamjsonrpc/blob/master/src/StreamJsonRpc/MessageHandlerBase.cs#L220). In the version of JsonRpc that's currently in VS this is called whenever connection is dropped. Newer version calls [`DisposeAsync`](https://github.com/microsoft/vs-streamjsonrpc/blob/46d47586cf611708174bd7fdb755cbbf6ce1d165/src/StreamJsonRpc/JsonRpc.cs#L2082-L2090), so presumably it shouldn't cause threads to hang anymore.

**Comment by CyrusNajmabadi** [comment]
Going to close out now that 16.5p2 is out.  Will open a new issue if i still run into things.
